### PR TITLE
[ssh] improvement

### PR DESCRIPTION
**Pull request by XiaoliChan**
Totally bugs fixed & improvement:

- `[ssh.py]`: less create `ssh connect`, keep doing set credential via `paramiko transport`
- `[ssh.py]`: rewrite `enum_host_info` function
- `[ssh.py]`: fix hanging, old one will never exit
- `[ssh.py]`: fix private key with passphrase
- `[ssh.py]`: add sudo check for linux user
- `[ssh.py]`: windows privileges check
- `[ssh.py]`: improve command execute and format command execute result
- `[ssh.py]`: `paramiko` always discovery private keys in `~/.ssh/`, that will make `paramiko` exception, disable it.

### CODE REVIEW

File: nxc/protocols/ssh.py
Comment (Marshall-Hallenbeck) [inline nxc/protocols/ssh.py:49]: I think we should just set the version to "Unknown", since the variable is remote version, we already know what it is.

File: nxc/protocols/ssh.py
Comment (Marshall-Hallenbeck) [inline nxc/protocols/ssh.py:83]: I don't think we should use && to check for this - we should run it one by one like previously.

File: nxc/protocols/ssh.py
Comment (Marshall-Hallenbeck) [inline nxc/protocols/ssh.py:85]: rename "admin_flag"

File: nxc/protocols/ssh.py
Comment (Marshall-Hallenbeck) [inline nxc/protocols/ssh.py:116]: How exactly is this checking for sudo, other than the piping into sudo?
It's also an opsec concern, since we are passing passwords on the command line.

File: nxc/protocols/ssh.py
Comment (Marshall-Hallenbeck) [inline nxc/protocols/ssh.py:143]: Opsec concern - we might want to make this and the previous check modules

File: nxc/protocols/ssh.py
Comment (Marshall-Hallenbeck) [inline nxc/protocols/ssh.py:219]: Which IoT devices?

File: nxc/protocols/ssh.py
Comment (Marshall-Hallenbeck) [inline nxc/protocols/ssh.py:246]: This is making an assumption... is there anyway for us to determine the specific OS?

File: nxc/protocols/ssh/proto_args.py
Comment (Marshall-Hallenbeck) [inline nxc/protocols/ssh/proto_args.py:8]: Can you use an f-string here?

File: pyproject.toml
Comment (Marshall-Hallenbeck) [inline pyproject.toml:45]: What's needed in the new version?

**Review by Marshall-Hallenbeck — CHANGES_REQUESTED** [review]
Will need to actually test this as well

@XiaoliChan can you also add new e2e tests in the command list for the new options you've added?

File: pyproject.toml
Comment (XiaoliChan) [reply pyproject.toml:45]: New version has better performance and lots of bugs fixed

File: nxc/protocols/ssh/proto_args.py
Comment (XiaoliChan) [reply nxc/protocols/ssh/proto_args.py:8]: Argparser not allow f-string

File: nxc/protocols/ssh.py
Comment (XiaoliChan) [reply nxc/protocols/ssh.py:246]: I don't think it can, because different network device like Cisco, Huawei, Netgear has different ssh shell prompt, it's really hard to detect network device OS version

File: nxc/protocols/ssh.py
Comment (XiaoliChan) [reply nxc/protocols/ssh.py:219]: I test it with my home network switch (Netgear GS308T)

File: nxc/protocols/ssh.py
Comment (XiaoliChan) [reply nxc/protocols/ssh.py:83]: Yes

**Comment by XiaoliChan** [comment]
> Will need to actually test this as well
> 
> @XiaoliChan can you also add new e2e tests in the command list for the new options you've added?

Sure!

File: nxc/protocols/ssh.py
Comment (NeffIsBack) [inline nxc/protocols/ssh.py:23]: The Debug statement from the original connection.py would be nice to have here

File: nxc/protocols/ssh.py
Comment (NeffIsBack) [inline nxc/protocols/ssh.py:59]: I think these are better placed in the `proto_logger` function imo

File: nxc/protocols/ssh.py
Comment (NeffIsBack) [inline nxc/protocols/ssh.py:102]: I think this one could be reduced to simply:
```python
if not self.admin_privs:
    tips = admin_Flag[match[0]][1]
```

If i am not mistaken

File: nxc/protocols/ssh.py
Comment (NeffIsBack) [inline nxc/protocols/ssh.py:188]: This should be placed after the login with username&password to ensure only valid credentials are added to the db

File: nxc/protocols/ssh.py
Comment (NeffIsBack) [inline nxc/protocols/ssh.py:269]: Why removing the password from the shell output?

File: nxc/protocols/ssh.py
Comment (NeffIsBack) [inline nxc/protocols/ssh.py:242]: Although this is working out in the end with the "non" in line 271 if `self.admin_privs` is false, only seeing these two lines is really confusing :D

I would prefer adding the "non admin" here in this line rather than 271 if that is okay :)

**Review by NeffIsBack — CHANGES_REQUESTED** [review]

**Comment by NeffIsBack** [comment]
image
image

Linux works 👍

**Comment by XiaoliChan** [comment]
@NeffIsBack @Marshall-Hallenbeck I will solve the issues in reviews after my holiday :D

**Comment by NeffIsBack** [comment]
> @NeffIsBack @Marshall-Hallenbeck I will solve the issues in reviews after my holiday :D

Take your time and enjoy your holidays :D

File: nxc/protocols/ssh.py
Comment (XiaoliChan) [reply nxc/protocols/ssh.py:83]: Ha, I think ";" is better

File: nxc/protocols/ssh.py
Comment (XiaoliChan) [reply nxc/protocols/ssh.py:116]: Yes, this is doing something like pipe

File: nxc/protocols/ssh.py
Comment (XiaoliChan) [reply nxc/protocols/ssh.py]: Fixed!

File: nxc/protocols/ssh.py
Comment (XiaoliChan) [reply nxc/protocols/ssh.py:270]: I think ‘{password}’ is useless here, cause we use private key file

File: nxc/protocols/ssh.py
Comment (XiaoliChan) [reply nxc/protocols/ssh.py]: Fixed

File: nxc/protocols/ssh.py
Comment (XiaoliChan) [reply nxc/protocols/ssh.py:28]: Fixed

File: nxc/protocols/ssh.py
Comment (XiaoliChan) [reply nxc/protocols/ssh.py]: Fixed

File: nxc/protocols/ssh.py
Comment (XiaoliChan) [reply nxc/protocols/ssh.py]: Fixed

File: nxc/protocols/ssh.py
Comment (XiaoliChan) [reply nxc/protocols/ssh.py]: Fixed

File: nxc/protocols/ssh.py
Comment (XiaoliChan) [reply nxc/protocols/ssh.py:247]: Fixed

**Comment by XiaoliChan** [comment]
> Tho it would be nice if you could change the syntax back to for example "(Pwn3d!) - (root/non root) Linux - Shell access!"

Done, I changed it to this
image

**Comment by XiaoliChan** [comment]
Confirm the private key with passphrase login works
image

File: pyproject.toml
Comment (NeffIsBack) [reply pyproject.toml:45]: Is there a reason to pin the version to 3.3.1? Why not `^3.3.1`

Comment (XiaoliChan) [reply pyproject.toml:45]: I'm not good at using poetry stuff XD, it's better to use `^3.3.1`?

Comment (NeffIsBack) [reply pyproject.toml:45]: Haha no problem. With `=` you pin the installed version to exactly the specified version. Using `^` specifies that installed version must atleast match the specified version but allows updates with `poetry update` to higher versions except a major version bump. (The exact version that is installed then is specified in the lock file)
So probably we should use `^`.

See: https://python-poetry.org/docs/dependency-specification/

Comment (XiaoliChan) [reply pyproject.toml:45]: Ha, I see, thanks you!

File: nxc/protocols/ssh.py
Comment (XiaoliChan) [reply nxc/protocols/ssh.py]: Solved!

File: nxc/protocols/ssh.py
Comment (XiaoliChan) [reply nxc/protocols/ssh.py]: Like this?

Comment (XiaoliChan) [reply nxc/protocols/ssh.py]: Hmm, sorry for my comprehension

File: nxc/protocols/ssh/proto_args.py
Comment (NeffIsBack) [inline nxc/protocols/ssh/proto_args.py:7]: Please make `--sudo-check-method` require `--sudo-check`, otherwise it won't make much sense. See ldap proto-args `--no-smb` requires `--domain` as an example

File: nxc/protocols/ssh.py
Comment (NeffIsBack) [inline nxc/protocols/ssh.py:128]: Please remove the `stdin, stdout, stderr` if we never read the actual output from the command. I think this also applies to other places in this file.

File: nxc/protocols/ssh.py
Comment (NeffIsBack) [inline nxc/protocols/ssh.py:132]: If you only need `stderr` for example like in this case please "throw away" the other variables with `_`, this increases readability a lot because you know now what to focus on (and is a bit more efficient i think). Here i think it would be `_, _, stderr = self.con...`

**Review by NeffIsBack — CHANGES_REQUESTED** [review]

**Comment by NeffIsBack** [comment]
image

Windows working for me 🎉

File: nxc/protocols/ssh.py
Comment (NeffIsBack) [inline nxc/protocols/ssh.py:199]: Please also make sure to use `'` only inside strings. Outside quotes should always be double quotes `"`

File: nxc/protocols/ssh/proto_args.py
Comment (XiaoliChan) [reply nxc/protocols/ssh/proto_args.py:7]: Okay :)

File: nxc/protocols/ssh.py
Comment (XiaoliChan) [reply nxc/protocols/ssh.py:133]: Yes yes, you're right, should paste for anonymous variable

File: nxc/protocols/ssh.py
Comment (XiaoliChan) [reply nxc/protocols/ssh.py:199]: Sure!

File: nxc/protocols/ssh.py
Comment (XiaoliChan) [reply nxc/protocols/ssh.py]: Fixed!

File: nxc/protocols/ssh.py
Comment (XiaoliChan) [reply nxc/protocols/ssh.py]: Fixed!

**Comment by NeffIsBack** [comment]
Code looks good, after testing this is ready for merge

File: nxc/protocols/ssh.py
Comment (Marshall-Hallenbeck) [inline nxc/protocols/ssh.py:115]: Just an English fix (and remove the period): "Check admin with sudo does not support using a private key"

File: nxc/protocols/ssh.py
Comment (Marshall-Hallenbeck) [inline nxc/protocols/ssh.py:127]: should be `shadow_backup`, no upper case B

File: nxc/protocols/ssh.py
Comment (Marshall-Hallenbeck) [inline nxc/protocols/ssh.py:136]: Should be f-string and recommended English fix:
`self.logger.info(f"The file {shadow_backup} does not exist, the pipe may be hanging. Increase the number of tries with the option "--get-output-tries" or change other method with "--sudo-check-method". If it's still failing, maybe sudo shell does not work with the current user")`

File: nxc/protocols/ssh.py
Comment (Marshall-Hallenbeck) [inline nxc/protocols/ssh.py:138]: Change to double quotes

File: nxc/protocols/ssh.py
Comment (Marshall-Hallenbeck) [inline nxc/protocols/ssh.py:158]: change to `shadow_Backup`

File: nxc/protocols/ssh.py
Comment (Marshall-Hallenbeck) [inline nxc/protocols/ssh.py:172]: Apply same fix as the other area

File: nxc/protocols/ssh.py
Comment (Marshall-Hallenbeck) [inline nxc/protocols/ssh.py:170]: Change to double quotes

File: nxc/protocols/ssh/proto_args.py
Comment (Marshall-Hallenbeck) [inline nxc/protocols/ssh/proto_args.py:16]: Just use single quotes here inside the string instead of escaping double quotes

**Review by Marshall-Hallenbeck — CHANGES_REQUESTED** [review]
Thank you for adding the new commands to the tests too 😊

File: nxc/protocols/ssh.py
Comment (XiaoliChan) [reply nxc/protocols/ssh.py:115]: Fixed!

File: nxc/protocols/ssh.py
Comment (XiaoliChan) [reply nxc/protocols/ssh.py:127]: Fixed!

File: nxc/protocols/ssh.py
Comment (XiaoliChan) [reply nxc/protocols/ssh.py:136]: Fixed!

File: nxc/protocols/ssh.py
Comment (XiaoliChan) [reply nxc/protocols/ssh.py:138]: Fixed!

File: nxc/protocols/ssh.py
Comment (XiaoliChan) [reply nxc/protocols/ssh.py:158]: Fixed!

File: nxc/protocols/ssh.py
Comment (XiaoliChan) [reply nxc/protocols/ssh.py:172]: Fixed!

File: nxc/protocols/ssh.py
Comment (XiaoliChan) [reply nxc/protocols/ssh.py:170]: Fixed!

File: nxc/protocols/ssh/proto_args.py
Comment (XiaoliChan) [reply nxc/protocols/ssh/proto_args.py:16]: Fixed!

File: nxc/protocols/winrm/proto_args.py
Comment (XiaoliChan) [inline nxc/protocols/winrm/proto_args.py:1]: Wrong fix. XD

**Comment by NeffIsBack** [comment]
There are still some single quotes instead of double quotes :)

**Review by NeffIsBack — APPROVED** [review]
image

Everything is working now. Great job @XiaoliChan 

There are still a few single quotes but we wait until #35 is merged, probably a lot are already fixed there

**Review by NeffIsBack — CHANGES_REQUESTED** [review]
image

Have to revoke approval, ssh with keyfile is no longer working from linux to linux

**Comment by NeffIsBack** [comment]
> Have to revoke approval, ssh with keyfile is no longer working from linux to linux

Further investigation:
having ssh keyfiles in the `~/.ssh` directory leads ssh to crash
image

**Comment by XiaoliChan** [comment]
> Further investigation: having ssh keyfiles in the `~/.ssh` directory leads ssh to crash image

Fixed!

**Comment by NeffIsBack** [comment]
image

image

Everything working now 👍

**Review by NeffIsBack — APPROVED** [review]

**Comment by NeffIsBack** [comment]
@Marshall-Hallenbeck this is ready as well, should i merge it tho or are we waiting for your PR?

**Comment by Marshall-Hallenbeck** [comment]
@NeffIsBack we can probably merge this, I still have to fix some of the things from your feedback for #35

**Review by Marshall-Hallenbeck — APPROVED** [review]

**Pull request merged**
